### PR TITLE
Y24-095 Message processing (consuming) for Volume Tracking messages from Traction Service

### PR DESCRIPTION
**Pull request by dasunpubudumal**
Closes #

#### Changes proposed in this pull request

- 

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: tol_lab_share/messages/mlwh/create_aliquot_message.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            "volume": self.volume,
+            "concentration": self.concentration,
+            "insert_size": self.insert_size,
+            "last_updated": self.last_updated,
+            "recorded_at": self.recorded_at,
+            "created_at": self.created_at,
+        }
+
+
+class CreateAliquotInWarehouseMessage(MessageProperty):
```
Comment (sdjmchattie) [inline tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: I understand why you followed the pattern with the `MessageProperty` class, but I really think this is a pattern I wanted to step away from.  We shouldn't be doing such extensive validation of the values in the message when we're in full control of the message creation.  In almost every case, Avro schemas provide us with all the validation we need that values are of the correct type.  I appreciate that we might not have a schema defined for the message going into the MLWH RabbitMQ, but perhaps we should have?

File: tol_lab_share/config/rabbit.py [hunk 2]
```diff
@@ -20,6 +22,20 @@
     vhost="tol",
 )
 
+# In our servers, this will be picked up using deployment project's
+# roles/deploy_tol_stack/templates/tol-lab-share/app_config.py.j2 and
+# environments/uat/group_vars/tol_swarm_managers.yml
+ENVIRONMENT_NAME = os.environ.get("ENVIRONMENT_NAME", "development")
```
Comment (sdjmchattie) [inline tol_lab_share/config/rabbit.py:28]: This seems to be quite specific to MLWH.  Should it be named more appropriately?

File: tol_lab_share/constants/__init__.py [hunk 3]
```diff
@@ -2,11 +2,13 @@
 
 
 RABBITMQ_SUBJECT_BIOSCAN_POOL_XP_TO_TRACTION = "bioscan-pool-xp-tube-to-traction"
+RABBITMQ_SUBJECT_CREATE_ALIQUOT_IN_MLWH = "create-aliquot-in-mlwh"
 RABBITMQ_SUBJECT_CREATE_LABWARE = "create-labware"
 RABBITMQ_SUBJECT_UPDATE_LABWARE = "update-labware"
 RABBITMQ_FEEDBACK_EXCHANGE = "psd.tol"
 RABBITMQ_ROUTING_KEY_CREATE_LABWARE_FEEDBACK = "feedback.created.labware"
 RABBITMQ_ROUTING_KEY_UPDATE_LABWARE_FEEDBACK = "feedback.updated.labware"
+RABBITMQ_ROUTING_KEY_CREATE_ALIQUOT = "{}.saved.aliquot.{}"
```
Comment (sdjmchattie) [inline tol_lab_share/constants/__init__.py:11]: If this is necessary, can we use named formatting parameters?

```
txt = "{environment}.saved.aliquot.{lims_uuid}"
txt.format(environment = "development", lims_uuid = "mlwh")
```

**Review by sdjmchattie — CHANGES_REQUESTED** [review]
Mostly naming suggestions, but I do think we should consider whether there's an easier way to generate the messages going to MLWH, at most only using an Avro schema to validate the contents.  The existing pattern is way too hard to follow and to appreciate what the message contents might be.

File: tol_lab_share/messages/mlwh/create_aliquot_message.py (hunk 1 above)
Comment (dasunpubudumal) [reply tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: I actually agree, as I had the same thoughts on the validations; I realised it when I was writing the tests.

File: tol_lab_share/processors/create_aliquot_processor.py [hunk 4]
```diff
@@ -32,13 +32,35 @@ def __init__(self, schema_registry: SchemaRegistry, basic_publisher: BasicPublis
 
     @staticmethod
     def instantiate(
-        schema_registry: SchemaRegistry, basic_publisher: BasicPublisher, config: Any
+        _schema_registry: SchemaRegistry, _basic_publisher: BasicPublisher, config: Any
```
[the remark this answers is not in the document]
Comment (dasunpubudumal) [reply tol_lab_share/processors/create_aliquot_processor.py]: mm weird; I don't remember doing this. 

Anyway, thanks. I'll update it.

File: tol_lab_share/messages/mappers/traction_to_warehouse.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import logging
+
+from tol_lab_share.constants.input_traction_volume_tracking_message import TRACTION_LIMS
+from tol_lab_share.messages.mlwh.create_aliquot_message import CreateAliquotInWarehouseMessage
+from tol_lab_share.messages.consumed import TractionToWarehouseMessage
+
+logger = logging.getLogger(__name__)
+
+
+class TractionToWarehouseMapper:
```
Comment (sdjmchattie) [inline tol_lab_share/messages/mappers/traction_to_warehouse.py:11]: Sorry I should have also commented about the name of this class.  We should make it more specific to the use of this message to avoid future naming complications.

File: tol_lab_share/messages/mlwh/create_aliquot_message.py (hunk 1 above)
Comment (dasunpubudumal) [reply tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: And I suppose it's not _that_ difficult to get the schema in and do the validations at least on the producer side (`tol-lab-share` in this case). I'll put this into the conversation tomorrow.

Comment (sdjmchattie) [reply tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: When I say using Avro, I do only mean on the publisher here.  We needn't modify how MLWH works as long as it can support the JSON encoded format of Avro.  If that's not the case, don't bother going down this route and do traditional type validation.  JSON format isn't highly supported across all libraries for Avro, but `fastavro` as used here is very good at it.

Comment (dasunpubudumal) [reply tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: mm well, the warehouse is a thin Rails application (there's technical term for such Rails applications; I can't remember it!) that has the queue as some sort of a sidecar so if it supports JSON formatting I guess there's no harm in using it. And I think we have integrating schema validations at the warehouse side in our roadmap if I remember our discussions before correctly.

Comment (dasunpubudumal) [reply tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: @sdjmchattie well, maybe I am looking in the wrong documentation, but [`avro` gem](https://www.rubydoc.info/gems/avro/1.8.1) in Ruby does support binary encoding and decoding. It doesn't say anything about dealing with a JSON-encoded format. For instance, when I JSON-encode (with [this schema](https://redpanda.uat.psd.sanger.ac.uk/console/schema-registry/subjects/create-aliquot-in-mlwh?version=latest)) the following message using `fastavro`:

```json
{
  "messageUuid": "b66d5daa-f1d5-4a74-9e93-969971043a42",
  "messageCreateDateUtc": 1578836705000,
  "limsId": "LIMS123",
  "limsUuid": "a8e4a3c3-954a-445b-b92f-bcec84029a3d",
  "aliquotType": "derived",
  "sourceType": "sample",
  "sourceBarcode": "SRC123456789",
  "sampleName": "Sample123",
  "usedByType": "library",
  "usedByBarcode": "LIB123456789",
  "volume": 150.5,
  "concentration": 20.0,
  "insertSize": 500,
  "recordedAt": 1578836705000
}
```

I get the following:

```json
{
  "messageUuid": "b66d5daa-f1d5-4a74-9e93-969971043a42",
  "messageCreateDateUtc": 1578836705000,
  "limsId": "LIMS123",
  "limsUuid": "a8e4a3c3-954a-445b-b92f-bcec84029a3d",
  "aliquotType": "derived",
  "sourceType": "sample",
  "sourceBarcode": "SRC123456789",
  "sampleName": "Sample123",
  "usedByType": "library",
  "usedByBarcode": "LIB123456789",
  "volume": 150.5,
  "concentration": {
    "float": 20
  },
  "insertSize": {
    "int": 500
  },
  "recordedAt": 1578836705000
}
```

I guess `avro` gem requires the ability to decode a message like above if we are going to json-encode it in tls? The warehouse expects the message to be of the structure similar to the message I encoded.

Comment (sdjmchattie) [reply tol_lab_share/messages/mlwh/create_aliquot_message.py:56]: The Ruby gem does not support JSON encoding or decoding.  My assumption is that the message on the MLWH side will not be changed and we won't be using Avro there.  I see that union types show the type in the JSON format for Avro.  I didn't know that and I don't think this will allow it to be suitable for creating the MLWH message.

File: SETUP_DOCKER.md [hunk 6]
```diff
@@ -30,12 +30,18 @@ to run this one as it requires Rabbitmq to have started completely:
...
+    RABBITMQ_PORT=5672
+    RABBITMQ_USERNAME=admin
+    RABBITMQ_PASSWORD=development
     ```
 
 1. Start interactive bash in docker container
 
     ```bash
-    docker run -ti -v $(pwd):/code --env-file=.env --entrypoint bash tol-lab-share:develop
+   docker run -ti -v $(pwd):/code -v /etc/ssl/certs/ca-certificates.crt:/etc/ssl/certs/ca-certificates.crt:ro --env-file=.env --entrypoint bash tol-lab-share:develop
```
[the remark this answers is not in the document]
Comment (sdjmchattie) [reply SETUP_DOCKER.md:44]:
```suggestion
    docker run -ti -v $(pwd):/code -v /etc/ssl/certs/ca-certificates.crt:/etc/ssl/certs/ca-certificates.crt:ro --env-file=.env --entrypoint bash tol-lab-share:develop
```

File: schemas/create-aliquot-in-mlwh/1-create-aliquot-in-mlwh.avsc [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            "name": "recordedAt",
+            "type": {
+                "type": "long",
+                "logicalType": "timestamp-millis"
+            },
+            "doc": "The date and time that the aliquot was recorded"
+        }
+    ]
+}
```
[the remark this answers is not in the document]
Comment (sdjmchattie) [reply schemas/create-aliquot-in-mlwh/1-create-aliquot-in-mlwh.avsc:124]:
```suggestion
}

```

File: tol_lab_share/messages/consumed/traction_to_warehouse/traction_to_mlwh_aliquot.py
Comment (sdjmchattie) [inline tol_lab_share/messages/consumed/traction_to_warehouse/traction_to_mlwh_aliquot.py:1]: I've just been comparing this to the bioscan implementation and I wonder if this should be in a folder called `traction_to_mlwh_aliquot` and the filename be `root.py` to match.

Comment (dasunpubudumal) [reply tol_lab_share/messages/consumed/traction_to_warehouse/traction_to_mlwh_aliquot.py:1]: yeah makes sense.

File: schemas/create-aliquot-in-mlwh/1-create-aliquot-in-mlwh.avsc [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            "type": {
+                "type": "enum",
+                "name": "allowedUsedByType",
+                "symbols": [
+                    "library",
+                    "pool",
+                    "run",
+                    "none"
+                ],
+                "default": "none"
```
[the remark this answers is not in the document]
Comment (dasunpubudumal) [reply schemas/create-aliquot-in-mlwh/1-create-aliquot-in-mlwh.avsc:83]: No worries! 

I think @seenanair did some tweaks to the schema based on a discussion. I cannot guarantee that I know 100% of what "none" stands for but I remember a quick discussion in the standup. Maybe @seenanair could give a hint?

[the remark this answers is not in the document]
Comment (seenanair) [reply schemas/create-aliquot-in-mlwh/1-create-aliquot-in-mlwh.avsc:83]: Aliquot creation is done in different contexts, out of which one of it is creating a 'primary' aliquot when a library is created. In this scenario, the aliquot is of type 'primary' and it will only have source and no used_by. 
Another context is when a library is used in a pool or run, we are creating a 'derived' aliquot. In this case, the aliquot is of type 'derived' and it will have source (library) and used_by (pool/run well).
So, it is not mandatory that used_by field is always filled up especially in first scenario.

I have made this change, after Stuart mentioned about an error where the  'none'  for 'usedByType' was not in the correct place to make it a default value.
`{
            "name": "usedByType",
            "type": {
                "type": "enum",
                "name": "allowedUsedByType",
                "symbols": [
                    "library",
                    "pool",
                    "run",
                    "none"
                ]
            },
            "default": "none",
            "doc": "The type of the entity that the aliquot is used by"
        },`

[the remark this answers is not in the document]
Comment (sdjmchattie) [reply schemas/create-aliquot-in-mlwh/1-create-aliquot-in-mlwh.avsc:83]: So we just need to make sure the schema here matches what is on RedPanda, if we even want to keep maintaining a second copy of schemas in this repo

File: tol_lab_share/error_codes.py [hunk 9]
```diff
@@ -75,7 +75,7 @@ def origin_for_json(self) -> Any:
             return "sample"
         return "root"
 
-    def json(self) -> dict[str, Any]:
+    def json(self) -> "dict[str, Any]":
```
Comment (sdjmchattie) [inline tol_lab_share/error_codes.py:78]: Out of interest, why was this necessary?  I know Python supports string descriptions of types, but it usually only needs this when you have a circular reference.

File: tools/demo/README.md [hunk 10]
```diff
@@ -1,6 +1,13 @@
 How to run
 ----------
 
+0. Spin up the required dependency containers:
```
Comment (sdjmchattie) [inline tools/demo/README.md:4]: I didn't even know you could start at 0?  It's usually wise in markdown files to number all the items in your list as `1.` from the off.  That way you don't end up in situations where adding items in the middle of the list causes you to have to refactor the whole list, or do what you did here.

**Review by sdjmchattie — APPROVED** [review]
Nothing major left.  Thanks for putting up with all my comments.  It would be good to understand why the type had to be quoted as I don't think it should be necessary.

File: tol_lab_share/error_codes.py (hunk 9 above)
Comment (dasunpubudumal) [reply tol_lab_share/error_codes.py:78]: I don't really think the quotes are required because we're working with Python 3.11 (which is > 3.9) but Pylance suggested me this.

image

File: tools/demo/README.md (hunk 10 above)
Comment (dasunpubudumal) [reply tools/demo/README.md:4]: Well I did what they did when naming the zeroth law of thermodynamics 😆 

I was going to refactor the list later and forgot. I'll refactor it 😄

File: tol_lab_share/error_codes.py (hunk 9 above)
Comment (sdjmchattie) [reply tol_lab_share/error_codes.py:78]: Hmmm.  Even if this were true, the correct fix is to use `Dict` instead of `dict` and import it from `typing` but that shouldn't be necessary as you said.
